packages/client/lib/execution/level.ts

```diff
@@ -20,16 +28,32 @@ export class LevelDB implements DB {
   constructor(
     leveldb?: AbstractLevel<string | Uint8Array, string | Uint8Array, string | Uint8Array>
   ) {
-    this._leveldb = leveldb ?? new MemoryLevel(ENCODING_OPTS)
+    this._leveldb = leveldb ?? new MemoryLevel()
   }
 
   /**
    * @inheritDoc
    */
-  async get(key: Uint8Array): Promise<Uint8Array | null> {
-    let value = null
+  // @ts-expect-error
+  async get(key: Uint8Array | string): Promise<Uint8Array | string | DBObject | undefined> {
+    let value
+    let encoding = undefined
+    // Set value encoding based on key type or specific key names so values are interpreted correctly by Level
+    if (
+      key instanceof Uint8Array ||
+      key === 'CliqueSigners' ||
+      key === 'CliqueVotes' ||
+      key === 'CliqueBlockSignersSnapshot'
+    )
+      encoding = 'view'
+    if (key === 'heads' || typeof key === 'number') {
+      encoding = 'json'
+    }
     try {
-      value = await this._leveldb.get(key, ENCODING_OPTS)
+      value = await this._leveldb.get(key, {
+        valueEncoding: encoding,
+      })
+      if (value === null) return undefined
```

Comment:
Ah, this is too hacky here, we need to find another solution for that. 🙂

(so I am getting this right that this would also be the "default" implementation for other Blockchain LevelDB users, right?)

My tendency would be that we make encoding a mandatory method parameter in the `DB` interface and then just always pass this explicitly in our implementations and one can then decide in the wrapper implementation s (like `LevelDB`, so this one here) to use this information or not.

Does this make sense?

Reply:
We can though I don't know if we should or not.  The nicest sort ofd "default" way would be that the DB just takes what we pass in without doing anything to it and returns that same value when we pass in the same key (more or less how our `MapDB` does).  The reason I have to do all this special handling is that `level` assumes that all data should be a certain datatype and format when you dump it into the DB.  I'm not an expert on key/value DBs but if other key/value DBs require encoding details and if the names that specify those encoding types is identical, I think it's fine.  If not, I'd prefer to err on the side of not overspecifying what should be in the wrapper.

Reply:
The wrapper is a code part which is not in our control. People will just copy over this code part if they want to use Blockchain with LevelDB. So this should have a very stable API and be robust and generic in the inside. We do not want to tell people "oh, and please add `if (key === NewKeyWeAddedToBlockchain)` in line 34 of your wrapper" when we publish an update of our Blockchain library where we add some new functionality. 🙂